src/ipa_pc/mod.rs

```diff
@@ -1204,146 +1223,214 @@ impl<G: AffineCurve, D: Digest> PolynomialCommitment<G::ScalarField> for InnerPr
         let lambda: G::ScalarField = fs_rng.squeeze_128_bits_challenge();
         let mut cur_challenge = G::ScalarField::one();
 
-        let poly_map: BTreeMap<_, _> = labeled_polynomials
-            .iter()
-            .map(|poly| (poly.label(), poly))
-            .collect();
-
         let mut has_hiding = false;
-        let mut points = vec![];
+
+        let h_poly_time = start_timer!(|| "Compute h(X) polynomial");
 
         // h(X)
-        let h_poly_time = start_timer!(|| "Compute batching polynomial");
-        let mut batch_polynomial = Polynomial::zero();
+        let mut h_polynomial = Polynomial::zero();
+
+        // z(X)
+        let mut z_polynomial = Polynomial::from_coefficients_slice(
+            &[G::ScalarField::one()]
+        );
 
         for (label, (_point_label, point)) in query_set.iter() {
             let labeled_polynomial =
-                poly_map.get(label).ok_or(Error::MissingPolynomial {
+                *poly_map.get(label).ok_or(Error::MissingPolynomial {
                     label: label.to_string(),
                 })?;
 
             if labeled_polynomial.hiding_bound().is_some() {
                 has_hiding = true;
             }
 
-            points.push(point);
-
             // y_i
-            let evaluated_y = labeled_polynomial.polynomial().evaluate(*point);
+            let y_i = labeled_polynomial.polynomial().evaluate(*point);
+
+            // (X - x_i)
+            let x_polynomial = Polynomial::from_coefficients_slice(&[
+                (G::ScalarField::zero() - point),
+                G::ScalarField::one()
+            ]);
+
+            // z(X) = PROD (X - x_i)
+            z_polynomial = &z_polynomial * &x_polynomial;
```

Comment:
does the standard polynomial multiplication use FFT? 
If so, we should change that to the "trivial", i.e. coefficient-wise, multiplication as these polynomials are of very low degree.

Reply:
I checked in ginger algebra/fft/...: It seems that if we define both polynomials as dense polynomials the multiplication is the FFT one. However, there is an implementation of the naive_mul, which would be the suitable one. Note that changing here to the naive mul will improve timings notably.

Reply:
Done